src/training_hub/algorithms/speculative_decoding.py

```
@@ -0,0 +1,1173 @@
+
+        if trust_remote_code:
+            cmd.append("--trust-remote-code")
+
+        # Set GPU visibility for managed vLLM subprocess
+        env = os.environ.copy()
+        if vllm_gpu_ids:
+            env["CUDA_VISIBLE_DEVICES"] = ",".join(str(g) for g in vllm_gpu_ids)
+            logger.info("vLLM GPUs: %s", env["CUDA_VISIBLE_DEVICES"])
+
+        logger.info("Launching vLLM server: %s", " ".join(cmd))
+        proc = subprocess.Popen(
```

Ruff (0.15.15)

[error] `subprocess` call: check for execution of untrusted input

(S603)